### PR TITLE
Cleanup singletons for 3 classes in ControllerStarter

### PR DESCRIPTION
**Pull request by jackjlli**
This PR cleans up singletons for `PinotLLCRealtimeSegmentManager`, `SegmentCompletionManager` and `RebalanceSegmentStrategyFactory`.

The benefits are: 
* `SegmentCompletionManager` should be injected instead of using a singleton pattern.
* `PinotLLCRealtimeSegmentManager` should be put inside HelixResourceManager (right now these two code are mixed together).
* `RebalanceSegmentStrategyFactory` is only used in `HelixResourceManager`, which shouldn't be a global object.
* Basically singleton shouldn't be created and explicitly freed.
* Singleton isn't the only way to identify there's only 1 object. Using instanceId to differentiate controller processes should be good enough.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java [hunk 1]
```diff
@@ -211,8 +214,8 @@ public void start() {
   private void setUpHelixController() {
     // Register and connect instance as Helix controller.
     LOGGER.info("Starting Helix controller");
-    _helixControllerManager = HelixSetupUtils
-        .setup(_helixClusterName, _helixZkURL, _instanceId, _isUpdateStateModel, _enableBatchMessageMode);
+    _helixControllerManager = HelixSetupUtils.setup(_helixClusterName, _helixZkURL, _instanceId, _isUpdateStateModel,
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java]: Have you changed your setting for the code format? Seems this part is reformatted back and forth.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java [hunk 2]
```diff
@@ -254,8 +257,15 @@ private void setUpPinotController() {
...
     // Helix resource manager must be started in order to create PinotLLCRealtimeSegmentManager
     LOGGER.info("Starting realtime segment manager");
-    PinotLLCRealtimeSegmentManager
-        .create(_helixResourceManager, _config, _controllerMetrics, _controllerLeadershipManager);
+
+    _pinotLLCRealtimeSegmentManager =
+        new PinotLLCRealtimeSegmentManager(_helixResourceManager, _config, _controllerMetrics,
+            _controllerLeadershipManager);
+    // TODO: Need to put this inside HelixResourceManager when ControllerLeadershipManager is removed.
+    _helixResourceManager.registerPinotLLCRealtimeSegmentManager(_pinotLLCRealtimeSegmentManager);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java]: This line can be moved into the constructor of PinotLLCRealtimeSegmentManager?

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java [hunk 3]
```diff
@@ -265,8 +275,9 @@ private void setUpPinotController() {
     _controllerPeriodicTaskScheduler = new ControllerPeriodicTaskScheduler();
     _controllerPeriodicTaskScheduler.init(controllerPeriodicTasks, _controllerLeadershipManager);
 
-    LOGGER.info("Creating rebalance segments factory");
-    RebalanceSegmentStrategyFactory.createInstance(helixParticipantManager);
+    LOGGER.info("Registering rebalance segments factory");
+    _helixResourceManager.registerRebalanceSegmentStrategyFactory(
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java]: This line can be moved into the constructor of RebalanceSegmentStrategyFactory and pass in _helixResourceManager

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 4]
```diff
@@ -139,25 +136,13 @@ public String getControllerVipUrl() {
     return _controllerConf.generateVipUrl();
   }
 
-  public static synchronized void create(PinotHelixResourceManager helixResourceManager, ControllerConf controllerConf,
+  public PinotLLCRealtimeSegmentManager(PinotHelixResourceManager helixResourceManager, ControllerConf controllerConf,
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: I think you can remove the old constructor and only keep this one. (For the test, use mock)

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/SegmentCompletionManager.java [hunk 5]
```diff
@@ -84,12 +81,15 @@ public static int getMaxCommitTimeForAllSegmentsSeconds() {
 
   // TODO keep some history of past committed segments so that we can avoid looking up PROPERTYSTORE if some server comes in late.
 
-  protected SegmentCompletionManager(HelixManager helixManager, PinotLLCRealtimeSegmentManager segmentManager,
-      ControllerMetrics controllerMetrics, ControllerLeadershipManager controllerLeadershipManager) {
+  public SegmentCompletionManager(HelixManager helixManager, PinotLLCRealtimeSegmentManager segmentManager,
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: Re-order the arguments probably?

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java (hunk 2 above)
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java]: I've tried that but since the object isn't fully initialized, it will only assign a null reference to it.
Plus, the logic of these 2 objects are mixed together: 
* `Resource manager` has to be instantiated before `PinotLLCRealtimeSegmentManager`.
* `Resource manager` needs to call the method in `PinotLLCRealtimeSegmentManager`, which is created after `Resource manager` got created.

It's a circular reference dependency between these two objects.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java (hunk 3 above)
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java]: Same as this one.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerStarter.java]: Make sense lol

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/SegmentCompletionManager.java (hunk 5 above)
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: Done.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java (hunk 4 above)
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: Done.
